packages/itwinui-react/src/utils/components/OverflowContainer.tsx

```diff
@@ -96,6 +157,263 @@ if (process.env.NODE_ENV === 'development') {
+ * @returns [callback ref to set on container, stateful count of visible items]
+ *
+ * @example
+ * const items = Array(10).fill().map((_, i) => <span>Item {i}</span>);
+ * const [ref, visibleCount] = useOverflow(items.count);
+ * ...
+ * return (
+ *   <div ref={ref}>
+ *     {items.slice(0, visibleCount)}
+ *   </div>
+ * );
+ */
+const useOverflow = (
+  itemsCount: number,
+  orientation: 'horizontal' | 'vertical' = 'horizontal',
+) => {
+  const [guessState, dispatch] = React.useReducer(
+    overflowGuessReducer,
+    { itemsCount },
+    overflowGuessReducerInitialState,
+  );
+
+  const { minGuess, maxGuess, isStabilized, visibleCount } = guessState;
+
+  const containerRef = React.useRef(null);
+  const isGuessing = React.useRef(false);
+
+  /**
+   * Call this function to guess the new `visibleCount`.
+   * The `visibleCount` is not changed if the correct `visibleCount` has already been found.
+   *
+   * Logic:
+   * - Have a guess range for `visibleCount`. e.g. `[0, 32]` (32 is an arbitrary choice).
+   * - Keep doubling the max guess until the container overflows.
+   *   i.e. the max guess should always be `≥` the correct `visibleCount`.
+   *   - With each such doubling, the new min guess is the current max guess (since underflow = we guessed low).
+   * - Set `visibleCount` to the `maxGuess`.
+   * - Repeat the following by calling `guessVisibleCount()` (keep re-rendering but not painting):
+   *   - Each time the container overflows, new max guess is the average of the two guesses.
+   *   - Each time the container does not overflow, new min guess is the average of the two guesses.
+   * - Stop when the average of the two guesses is the min guess itself. i.e. no more averaging possible.
+   * - The min guess is then the correct `visibleCount`.
+   */
+  const guessVisibleCount = React.useCallback(() => {
+    // If already stabilized, already guessing, or in unit test, do not guess.
+    if (isStabilized || isGuessing.current || isUnitTest) {
+      return;
+    }
+
+    try {
+      isGuessing.current = true;
+
+      // We need to wait for the ref to be attached so that we can measure available and required sizes.
+      if (containerRef.current == null) {
+        return;
+      }
+
+      // If there are no items, stabilize immediately.
+      if (itemsCount === 0) {
+        dispatch({ type: 'stabilize' });
+        return;
+      }
+
+      const dimension = orientation === 'horizontal' ? 'Width' : 'Height';
+      const availableSize = containerRef.current[`offset${dimension}`];
+      const requiredSize = containerRef.current[`scroll${dimension}`];
+
+      const isOverflowing = availableSize < requiredSize;
+
+      if (
+        // We have already found the correct visibleCount
+        (visibleCount === itemsCount && !isOverflowing) ||
+        // if the new average of visibleCountGuessRange will never change the visibleCount anymore (infinite loop)
+        (maxGuess - minGuess === 1 && visibleCount === minGuess)
```

Comment:
Could you look through all comments to make sure they're up-to-date? `visibleCountGuessRange` doesn't exist anymore, and some of the comments below might be in the wrong place, since the related code was moved into the reducer.